### PR TITLE
1310 valid pubbing

### PR DESCRIPTION
**Pull request by defong**
### Context
General validation of Pub

### Changes proposed in this pull request
Added Pub checking based on the csharp counterpart
mandatory fields, word count, numerical input

### Guidance to review

Clueless for nested enrichment validation.

borrowed concept from https://www.justinweiss.com/articles/a-lightweight-way-to-handle-different-validation-situations/

### CODE REVIEW

File: app/validators/words_count_validator.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+class WordsCountValidator < ActiveModel::EachValidator
```
Comment (dankmitchell) [inline app/validators/words_count_validator.rb]: I think rails has this out of the box i.e.

`validates_length_of : fee_details, :maximum => 250, :too_long => "Fee details can't be longetr than {{count}} words."), :tokenizer => lambda {|str| str.scan(/\w+/) }`

https://apidock.com/rails/ActiveRecord/Validations/ClassMethods/validates_length_of

File: app/models/course_enrichment.rb [hunk 2]
```diff
@@ -42,6 +42,18 @@ class CourseEnrichment < ApplicationRecord
 
   scope :latest_first, -> { order(created_at: :desc) }
 
+  validates :about_course, words_count: { max_words_count: 400 }, on: :publish
+  validates :interview_process, words_count: { max_words_count: 250 }, on: :publish
+  validates :how_school_placements_work, words_count: { max_words_count: 350 }, on: :publish
+
+  # salary vs fee needs forking
```
Comment (tvararu) [inline app/models/course_enrichment.rb]: Something like this?

```diff
-  # salary vs fee needs forking
	  validates :fee_international, :fee_uk_eu, numericality: { only_integer: true, greater_than_or_equal_to: 0, less_than_or_equal_to: 100000 }, on: :publish
-	  validates :fee_uk_eu, presence: true, on: :publish
-	  validates :salary_details, presence: true, on: :publish
	  validates :fee_details, words_count: { max_words_count: 250 }, on: :publish
	  validates :salary_details, words_count: { max_words_count: 250 }, on: :publish
	  validates :financial_support, words_count: { max_words_count: 250 }, on: :publish
	
+	  validate :fee_xor_salary
+	
+	  def fee_xor_salary
+	    fee_based = fee_uk_eu.present? || fee_international.present?
+	    salary_based = salary_details.present?
+	    unless fee_based ^ salary_based
+	      both_present = fee_based && salary_based
+	      errors.add(:base, "Specify either a fee or a salary#{both_present ? ", not both" : ""}")
+	    end
+	  end
```

File: app/validators/words_count_validator.rb (hunk 1 above)
Comment (defong) [reply app/validators/words_count_validator.rb]: nice one

File: spec/models/course_enrichment_spec.rb [hunk 3]
```diff
@@ -132,60 +132,91 @@
...
+              let(:course_enrichment) {
+                create(:course_enrichment,
+                  fee_uk_eu: presence,
+                  salary_details: presence,
+                  course: course)
+              }
+              it { should be false }
+
+              it 'add errors ' do
+                expect(course_enrichment.errors[:fee_uk_eu]).to match_array ["is not a number", "can't be blank"]
```
Comment (defong) [inline spec/models/course_enrichment_spec.rb]: struggling to do one liner test for these 

```              it { should validate_presence_of(:fee_uk_eu) }```

must be a context breakness

Comment (defong) [reply spec/models/course_enrichment_spec.rb]: Failure/Error: it { should validate_presence_of(:fee_uk_eu) }

     Shoulda::Matchers::ActiveModel::AllowValueMatcher::AttributeDoesNotExistError:
       The matcher attempted to set :fee_uk_eu on the FalseClass to nil, but
       that attribute does not exist.

File: app/models/course_enrichment.rb (hunk 2 above)
Comment (tvararu) [reply app/models/course_enrichment.rb]: Resolving as we've found a better solution offline.

File: app/validators/words_count_validator.rb (hunk 1 above)
Comment (defong) [reply app/validators/words_count_validator.rb]: oh dear
 Removed deprecated :tokenizer in the length validator 
https://github.com/rails/rails/commit/6a78e0ecd6122a6b1be9a95e6c4e21e10e429513

**Comment by defong** [comment]
> `validates_length_of` should help with max word count

Nope  that was for rails version 4.* and has been dropped for 5.*

**Comment by defong** [comment]
Ready for review again

File: spec/requests/api/v2/providers/courses/sync_with_search_and_compare_spec.rb [hunk 4]
```diff
@@ -84,5 +81,37 @@
...
+      context 'invalid enrichment' do
+        let(:invalid_enrichment) { create(:course_enrichment, :invalid_content) }
+
+        let(:course) { create(:course, :fee_type_based, provider: provider, enrichments: [invalid_enrichment]) }
+        it { should have_http_status(:unprocessable_entity) }
+
+        it 'has validation errors' do
+          expect(json_data.count).to eq 6
+          expect(response.body).to include('Invalid latest_enrichment')
+          expect(response.body).to include("Latest enrichment About course Reduce the word count for about course")
```
Comment (tvararu) [inline spec/requests/api/v2/providers/courses/sync_with_search_and_compare_spec.rb]: Hmm, I would be interested if we can clean these up and remove the humanised field name at the start, `About course Reduce ...` -> `Reduce ...`

Comment (tvararu) [reply spec/requests/api/v2/providers/courses/sync_with_search_and_compare_spec.rb]: https://stackoverflow.com/a/40254247

Comment (tvararu) [reply spec/requests/api/v2/providers/courses/sync_with_search_and_compare_spec.rb]: Okay, I'm happy to do this as a follow-up, will card it up. :)

Comment (tvararu) [reply spec/requests/api/v2/providers/courses/sync_with_search_and_compare_spec.rb]: This too: https://stackoverflow.com/a/43565674

**Comment by defong** [comment]
[Rebuild course publishing validations in Rails](https://trello.com/c/g7MtJU3T/1310-rebuild-course-publishing-validations-in-rails)

File: spec/requests/api/v2/providers/courses/sync_with_search_and_compare_spec.rb (hunk 4 above)
Comment (defong) [reply spec/requests/api/v2/providers/courses/sync_with_search_and_compare_spec.rb]: just fyi this test is just a little bit misleading, as I was clueless to what we should have it return

the `json_data`/`response.body` is structured like 

```ruby[{
    "title" => "Invalid latest_enrichment",
    "detail" =>
    "Latest enrichment About course Reduce the word count for about course",
    "source" => {}
  },
  {
    "title" => "Invalid latest_enrichment",
    "detail" =>
    "Latest enrichment Interview process Reduce the word count for interview process",
    "source" => {}
  },
  {
    "title" => "Invalid latest_enrichment",
    "detail" =>
    "Latest enrichment How school placements work Reduce the word count for how school placements work",
    "source" => {}
  },
  {
    "title" => "Invalid latest_enrichment",
    "detail" =>
    "Latest enrichment Fee international must be greater than or equal to 0",
    "source" => {}
  },
  {
    "title" => "Invalid latest_enrichment",
    "detail" => "Latest enrichment Fee uk eu must be less than or equal to 100000",
    "source" => {}
  },
  {
    "title" => "Invalid latest_enrichment",
    "detail" =>
    "Latest enrichment Fee details Reduce the word count for fee details",
    "source" => {}
]
```

Comment (tvararu) [reply spec/requests/api/v2/providers/courses/sync_with_search_and_compare_spec.rb]: It's okay I think

File: app/models/course.rb [hunk 5]
```diff
@@ -160,6 +177,10 @@ def funding
     end
   end
 
+  def is_fee_based?
+    self.funding == 'fee'
```
Comment (dankmitchell) [inline app/models/course.rb]: No need for `self` here

File: spec/factories/course_enrichments.rb [hunk 6]
```diff
@@ -53,8 +53,8 @@
         "Fees are made payable to the University.",
       ].sample
     end
-    fee_uk_eu         { fee_international * 2 }
-    fee_international { rand(100) * 10 }
+    fee_uk_eu         { Faker::Number.within(0..100000) }
```
Comment (dankmitchell) [inline spec/factories/course_enrichments.rb]: Great improvement in this file 💪

File: spec/factories/course_enrichments.rb [hunk 7]
```diff
@@ -98,4 +98,23 @@
     status { :draft }
     last_published_timestamp_utc { 5.days.ago }
   end
+
+  trait :invalid_content do
```
Comment (dankmitchell) [inline spec/factories/course_enrichments.rb]: Prehaps this should be more explicit i.e. `trait :invalid_enrichment do`

Comment (defong) [reply spec/factories/course_enrichments.rb]: let(:invalid_enrichment) { create(:course_enrichment, `:with_invalid_content`) }

how about `:with_invalid_content`?

File: spec/models/course/is_fee_based_spec.rb [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+      it { should be false }
+    end
+
+    context 'salary based course' do
+      let(:course) {
+        create(:course, :fee_type_based)
+      }
+
+      it { should be true }
```
Comment (dankmitchell) [inline spec/models/course/is_fee_based_spec.rb]: `its(: is_fee_based?) { should be_truthy }`

**Pull request merged**